### PR TITLE
clientv3: simplify watcher synchronization

### PR DESCRIPTION
**Pull request by heyitsanthony**
Was more complicated than it needed to be.

Also fixes clobbering id's on resume and losing watcher channels when watchers are disconnected and canceled.

/cc @hongchaodeng

### CODE REVIEW

**Comment by gyuho** [comment]
> Also fixes clobbering id's on

Do we need to backport this then?

**Comment by heyitsanthony** [comment]
@gyuho probably, it fixes the new test case

File: clientv3/watch.go [hunk 1]
```diff
@@ -284,6 +284,9 @@ func (w *watcher) Watch(ctx context.Context, key string, opts ...OpOption) Watch
 	if ok {
 		select {
 		case ret := <-retc:
+			if ret == nil {
```
Comment (gyuho) [inline clientv3/watch.go]: When does this happen? Can we add some comment? Thanks!

File: clientv3/watch.go [hunk 2]
```diff
@@ -118,8 +118,8 @@ type watchGrpcStream struct {
 
 	// mu protects the streams map
 	mu sync.RWMutex
```
Comment (xiang90) [inline clientv3/watch.go]: it seems that this mu also protects closeErr?

File: clientv3/watch.go [hunk 3]
```diff
@@ -118,8 +118,8 @@ type watchGrpcStream struct {
...
-	// streams holds all active watchers
-	streams map[int64]*watcherStream
+	// substreams holds all active watchers
```
Comment (xiang90) [inline clientv3/watch.go]: holds all active gRPC streams for watchers?

File: clientv3/watch.go [hunk 4]
```diff
@@ -369,9 +367,7 @@ func (w *watchGrpcStream) addStream(resp *pb.WatchResponse, pendingReq *watchReq
 		ws.initReq.rev = resp.Header.Revision
 	}
 
-	w.mu.Lock()
-	w.streams[ws.id] = ws
-	w.mu.Unlock()
+	w.substreams[ws.id] = ws
```
Comment (xiang90) [inline clientv3/watch.go]: why do we drop the lock here?

File: clientv3/watch.go [hunk 5]
```diff
@@ -380,45 +376,47 @@ func (w *watchGrpcStream) addStream(resp *pb.WatchResponse, pendingReq *watchReq
...
 	// close subscriber's channel
 	close(ws.outc)
-	delete(w.streams, ws.id)
-	empty := len(w.streams) == 0
-	if empty && w.stopc != nil {
-		w.stopc = nil
-	}
-	w.mu.Unlock()
-	return empty
+	delete(w.substreams, ws.id)
```
Comment (xiang90) [inline clientv3/watch.go]: why do we drop the lock here?

File: clientv3/watch.go [hunk 6]
```diff
@@ -380,45 +376,47 @@ func (w *watchGrpcStream) addStream(resp *pb.WatchResponse, pendingReq *watchReq
...
+			if _, ok := closing[ws]; !ok {
+				close(ws.recvc)
+			}
+		}
+		for len(w.substreams) != 0 {
+			ws := <-w.closingc
+			w.closeStream(ws)
+		}
+
+		w.owner.mu.Lock()
```
Comment (xiang90) [inline clientv3/watch.go]: probably add a delStream func on owner?

File: clientv3/watch.go [hunk 7]
```diff
@@ -520,9 +514,7 @@ func (w *watchGrpcStream) run() {
 
 // dispatchEvent sends a WatchResponse to the appropriate watcher stream
 func (w *watchGrpcStream) dispatchEvent(pbresp *pb.WatchResponse) bool {
-	w.mu.RLock()
-	defer w.mu.RUnlock()
-	ws, ok := w.streams[pbresp.WatchId]
+	ws, ok := w.substreams[pbresp.WatchId]
```
Comment (xiang90) [inline clientv3/watch.go]: why do we drop the lock here?

**Comment by xiang90** [comment]
@heyitsanthony OK. After read the whole thing, I understand that we want to serialize the closing event in the single run routine. Can we update the description of the mutex then?

File: clientv3/watch.go (hunk 3 above)
Comment (heyitsanthony) [inline clientv3/watch.go]: That's in `watcher.streams`, this tracks the watch id's that are in a single grpc stream (hence the rename from streams to substreams, using "streams" twice is super confusing). Will clarify the comments a bit.

**Comment by heyitsanthony** [comment]
there's still some intrinsic raceiness between the resume path and stream cancelation, going to rework this a little bit more so the substream goroutines don't have to reason about resume

**Comment by heyitsanthony** [comment]
Tore up the watcher code from the roots. The old code's resume path was totally broken and could livelock in some cases. Now, all watch registrations go through a `resuming` queue on `watchGrpcStream`. On reconnect, all `watcherStream` goroutines are stopped so the resuming revision is stable, then restarted and requeued on `resuming`. Overall the design is much more convincing than the old mess.

PTAL /cc @xiang90 @gyuho

File: clientv3/watch.go [hunk 8]
```diff
@@ -162,15 +163,18 @@ type watcherStream struct {
...
-	outc chan<- WatchResponse
+	outc chan WatchResponse
 	// recvc buffers watch responses before publishing
 	recvc chan *WatchResponse
-	id    int64
+	// donec closes when the watcherStream goroutine stops.
+	donec chan struct{}
+	// closing is set to true when stream should be scheduled to shutdown.
+	closing bool
+	// id is the registered watch id for on the grpc stream
```
Comment (gyuho) [inline clientv3/watch.go]: `s/for on/on/`?

File: clientv3/watch.go [hunk 9]
```diff
@@ -314,12 +320,7 @@ func (w *watcher) Close() (err error) {
 }
 
 func (w *watchGrpcStream) Close() (err error) {
-	w.mu.Lock()
```
Comment (gyuho) [inline clientv3/watch.go]: Q. Why do we remove this now? Was this causing any problem?

Comment (heyitsanthony) [inline clientv3/watch.go]: The sharing was making it unnecessarily difficult to reason about the `watcherStream` teardown path so now `watcherStream` will post itself to `watchGrpcStream.closingc` and `watchGrpcStream.run` handles the final removal of `watcherStream` resources.

File: clientv3/watch.go (hunk 8 above)
Comment (heyitsanthony) [inline clientv3/watch.go]: right, will fix

File: clientv3/watch.go [hunk 10]
```diff
@@ -561,140 +563,126 @@ func (w *watchGrpcStream) serveWatchClient(wc pb.Watch_WatchClient) {
...
+	if err != nil {
+		return nil, v3rpc.Error(err)
+	}
+	// mark all substreams as resuming
+	if len(w.substreams)+len(w.resuming) > 0 {
+		close(w.resumec)
+		w.resumec = make(chan struct{})
+		w.joinSubstreams()
+		for _, ws := range w.substreams {
+			ws.id = -1
```
Comment (xiang90) [inline clientv3/watch.go]: do we need to update the donec of wc as what we do for the resuming streams at line 654?

File: clientv3/watch.go [hunk 11]
```diff
@@ -490,43 +490,41 @@ func (w *watchGrpcStream) run() {
...
-				case <-w.donec:
-				}
-			}(pendingReq)
-			failedReq = nil
-			pendingReq = nil
+// nextResume chooses the next resuming to register with the grpc stream. Abandoned
+// streams are marked as nil in the queue since the head must wait for its inflight registration.
+func (w *watchGrpcStream) nextResume() *watcherStream {
+	for len(w.resuming) != 0 {
+		if w.resuming[0] != nil {
```
Comment (xiang90) [inline clientv3/watch.go]: is it possible to handle the nil case outside the next resume? it seems like we can always remove the abandoned resuming at the caller side, no?

File: clientv3/watch.go [hunk 12]
```diff
@@ -283,7 +288,10 @@ func (w *watcher) Watch(ctx context.Context, key string, opts ...OpOption) Watch
 	// receive channel
 	if ok {
 		select {
-		case ret := <-retc:
+		case ret, ok := <-wr.retc:
```
Comment (xiang90) [inline clientv3/watch.go]: can we add some comments for this? it is not clear to me why do we need to call watch recursively.

Comment (xiang90) [inline clientv3/watch.go]: i failed to figure out where we might close retc.

Comment (heyitsanthony) [inline clientv3/watch.go]: it doesn't any more; removed the close case

File: clientv3/watch.go (hunk 11 above)
Comment (heyitsanthony) [inline clientv3/watch.go]: It's possible, but the caller side would have to reimplement this loop every time. I'd rather have it done in one place.

File: clientv3/watch.go (hunk 10 above)
Comment (heyitsanthony) [inline clientv3/watch.go]: yes, but `ws` is appended into `resuming` and the following loop over `resuming` picks it up, so it's unnecessary here

File: clientv3/watch.go (hunk 11 above)
Comment (xiang90) [inline clientv3/watch.go]: OK

File: clientv3/watch.go (hunk 10 above)
Comment (xiang90) [inline clientv3/watch.go]: ah. right.

**Comment by xiang90** [comment]
LGTM. Fix CI?

**Pull request merged**